job/strategies/templates/strategy.py

```diff
@@ -19,39 +26,88 @@ class Strategy(metaclass=ABCMeta):
     popularity, semantic similarity).
     """
 
+    MAX_RECS = config.get("MAX_RECS")
+
     @abstractmethod
-    def fetch_data(self, site: Site, interactions_data: pd.DataFrame) -> List[Dict[str, Any]]:
+    def __init__(self, model_type: ModelType) -> None:
+        """
+        Constructor that needs to be defined by subclasses
+        """
+        self.model_type = model_type
+        self.experiment_time: datetime = datetime.now()
```

Comment:
🔧 I don't think this will allow different child strategies to have the exact same datetime, which we need for the sake of comparing apples to apples. Here's a minimal example:

```python
>>> class Foo(ABC):
...     def __init__(self):
...             self.dt = datetime.now()
... 
>>> class Bar(Foo):
...     def __init__(self):
...             super().__init__()
... 
>>> class Baz(Foo):
...     def __init__(self):
...             super().__init__()
... 
>>> Bar().dt
datetime.datetime(2022, 12, 22, 18, 6, 36, 598751)
>>> Baz().dt
datetime.datetime(2022, 12, 22, 18, 6, 42, 654601)
```

I'd advocate doing something similar to `Strategy.set_site`:

```python
def set_experiment_time(self, experiment_time: datetime) -> None:
    """TODO docstring"""
    self.experiment_time = experiment_time
```

And we can write another method that defines all these attributes that we couldn't define in `__init__` and that are necessary to the strategy's actual business logic, i.e., `fetch_data` and beyond to save time:

```python
class prepare(self, site: Site, experiment_time: datetime) -> None:
    """TODO docstring"""
    self.set_site(site)
    self.set_experiment_time(experiment_time)

    # Add other post-init but pre-fetch configurations & definitions as needed
```

Reply:
I knew that would not allow different child strategies to have the same exact `datetime`. But you answered my question: They **must** have the same `datetime` so you can compare different strategies.

Great. If that's the case, your solution is perfect. I really liked it.